design/Upgrade_Design.md

```diff
@@ -0,0 +1,225 @@
+Some upgrades could be consensus upgrades as well, for example, on a change of protocol.
+
+Note: Not critical for an upgrading mechanism to be present in phase 1.
+
+### The Wallet extension
+
+This is the component installed by end users which communicates with Obscuro nodes via RPC.
+For a good UX, Obscuro nodes must be backwards compatible and support even older versions of the Wallet extension.
+There must be warnings and mechanisms to help users upgrade to the latest version.
+
+Note: Not critical for an upgrading mechanism to be present in phase 1.
+
+## Attestation whitelisting
+
+The management contract will not support direct attestation verification, which means it has to whitelist a genesis enclave, 
+by relying on incentivising users and researchers to make sure that the attestation is the right code running on secure hardware.
+
+The genesis enclave for each new version must be published and whitelisted by the contract, together with a link to the source
+code that produces that binary image.
+
+This is the format of the event produced by the Management contract when the lifecycle is finalised.
+
+```
+ UpgradeEvent(
+   - WhitelistedGenesisAttestation // a node installed with the new version
+   - EnforcedHeight                // the rollup height from which only the new version will move forward. It can be -1 if not applicable.
+   - RotateMasterSeed              // boolean indicating whether 
+   - Attributes                    // a map of attributes that will be understood by the current version of the enclave.  
+   )
+```
+
+Note: In phase 1, the event can be emitted when the administrator of the contract calls an "approveVersion" function.
+This will give the upgrading power to the administrator, but in a transparent way.
+
+
+## Minimum viable upgrade mechanism
+
+This section describes the minimum implementation required in phase 1.
+
+Unlike most software that can get away with not thinking about upgrading during early development, TEE software doesn't have
+this luxury.
+Without a mechanism put in place in the initial release, the enclave cannot be upgraded at all because any other version will
+not be able to read the data.
+
+This is only a concern for the Obscuro enclave, as the upgrading of the Host component can be designed afterwards.
+
+### High-level tasks
+
+#### Management contract
+
+- Create the UpgradeEvent.
+- Add a function to publish metadata about a release.
+- Add a function callable only by the admin to approve one of the published releases.
+
+#### Enclave 
+
+1. Implement the UpgradeEvent authentication mechanism.
+   - Logic to understand the beacon chain canonical chain
+   - Minimal heuristics to mitigate "Weak subjectivity."
+
+2. Implement logic to verify an attestation against the approved version.
+
+3. Create an RPC endpoint to hand over secrets to an approved version, encrypted with the key.
+
+4. Logic to stop operating at the block height specified in the UpgradeEvent.
```

Comment:
Have you considered catastrophic bugs and potential recovery options?

I'm just wondering if there was ever a bug such that the state of the network meant that all enclaves crash and are not able to start-up on permitted versions then the data would be stuck. And unlike other networks, it might be very hard for accounts to prove their claims to funds trapped in the bridge because of the encryption. And because secrets have to be handed over I assume we need at least one node to be live to some extent to restore the rest.

Are there any break-glass solutions for those scenarios or do you consider those scenarios unrealistic?

Reply:
good point. Such a section is definitely needed